### PR TITLE
Added rig edit generator

### PR DESCRIPTION
**Pull request by SleepyEngi**
Creates rig edits automatically by posing the bones and then clicking "Generate rig edit" on the armature data tab.

If you want the models to be affected by the pose, remember to apply the armature modifiers first

### CODE REVIEW

File: __init__.py [hunk 1]
```diff
@@ -142,6 +142,16 @@ def execute(self, context):
         return handle_errors(self, shell.create_lods, {})
 
 
+class GenerateRigEdit(bpy.types.Operator):
+    """Generate rig edit nodes for all posed bones"""
+    bl_idname = "pose.generate_rig_edit"
```
Comment (HENDRIX-ZT2) [inline __init__.py:147]: Maybe more generic? Is there anything else we might add?

File: __init__.py [hunk 2]
```diff
@@ -214,6 +224,29 @@ def draw(self, context):
...
+        return True
+
+    def draw(self, context):
+        addon_updater_ops.check_for_update_background()
+        layout = self.layout
+        icon = preview_collection["frontier.png"].icon_id
+        row = layout.row(align=True)
+        sub = row.row()
+        sub.operator("pose.generate_rig_edit", icon_value=icon)
+        addon_updater_ops.update_notice_box_ui(self, context)
```
Comment (HENDRIX-ZT2) [inline __init__.py:247]: needed here?

File: plugin/utils/shell.py [hunk 3]
```diff
@@ -86,6 +87,183 @@ def create_lods():
 	return msgs
 
 
+def generate_rig_edit():
+    """Automatic rig edit generator by NDP. Detects posed bones and automatically generates nodes and offsets them."""
+    msgs = []
+    #logging.info(f"Generating rig edit")
+    print("-------------------------------------------------------------")
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:94]: Convert print to logging.info or appropriate log level

File: plugin/utils/shell.py [hunk 4]
```diff
@@ -86,6 +87,183 @@ def create_lods():
...
+    
+    #Check if selected object is a valid armature
+    if bpy.context.active_object.type != 'ARMATURE':
+        #Object is not an armature. Cancelling.
+        print("No armature selected.")
+        msgs.append(f"No armature selected.")
+        return msgs
+    
+    #Store current mode
+    Mode = bpy.context.mode
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:105]: always make normal variables lowercase only in python

File: plugin/utils/shell.py [hunk 5]
```diff
@@ -86,6 +87,183 @@ def create_lods():
...
+    armature = bpy.context.object
+    
+    #Iterate over every bone in the armature.
+    for i in armature.pose.bones:
+        #Check if the bones have been scaled
+        #Update locations
+        bpy.context.view_layer.update()
+        
+        #Check that no bones are scaled
+        if i.scale != Vector((1,1,1)):
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:123]: All those `!=` checks could reasonably fail for floating point math that is used to apply the transforms. Probably better to check with a helper function if each of the components math.is_close ? Maybe mathutils has a helper? idk

File: plugin/utils/shell.py [hunk 6]
```diff
@@ -86,6 +87,183 @@ def create_lods():
...
+            #Reset scale
+            #i.scale = Vector((1,1,1))
+            print("WARNING: "+str(i.name)+" HAD SCALE, CANCELLING")
+            msgs.append(f"Warning: {str(i.name)} had scale transforms. Reset scale for all bones and try again.")
+            return msgs
+        
+        #Find all bones with transforms.
+        if (i.location != Vector((0, 0, 0)) or i.rotation_quaternion != Quaternion((1,0,0,0)) or i.scale != Vector((1,1,1))) and i.name[:5] != "NODE_":
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:131]: more pythonic i.name.startswith("NODE_")

File: plugin/utils/shell.py [hunk 7]
```diff
@@ -86,6 +87,183 @@ def create_lods():
...
+    Edits = 0
+
+    #Force Switch to pose mode.
+    bpy.ops.object.mode_set(mode='POSE')
+    
+    #Get the armature
+    armature = bpy.context.object
+    
+    #Iterate over every bone in the armature.
+    for i in armature.pose.bones:
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:117]: Use better var name. A typical convention is `p_bone` for pose bones

File: plugin/utils/shell.py [hunk 8]
```diff
@@ -86,6 +87,183 @@ def create_lods():
...
+            #Switch to pose mode
+            bpy.ops.object.mode_set(mode='POSE')
+            
+            #Clear location
+            i.location = Vector((0,0,0))
+            #Clear rotation
+            i.rotation_quaternion = Quaternion((1,0,0,0))
+            
+            #Offset bonenode location
+            i.parent.matrix.translation = i.parent.matrix.translation + bonebase_pos_delt
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:206]: let's look at the matrix math later lol

**Review by HENDRIX-ZT2 — CHANGES_REQUESTED** [review]

File: __init__.py (hunk 1 above)
Comment (HENDRIX-ZT2) [reply __init__.py:147]: brain dead moment. I thought this was the panel, not the operator -.-

Comment (HENDRIX-ZT2) [reply __init__.py:147]: Maybe really call this 'Pose to Rig Edit'?

Comment (SleepyEngi) [reply __init__.py:147]: I don't know what else could require automation regarding rigs. Maybe for animations automatically fixing the IK setups. Could be "Generate Rig Edit Nodes"
I was thinking of adding options for
-apply armature modifiers of all armature children
-apply rest pose

File: __init__.py (hunk 2 above)
Comment (SleepyEngi) [reply __init__.py:247]: I don't know what this does but when I removed the row.row and layout.row lines the button disappeared

Comment (HENDRIX-ZT2) [reply __init__.py:247]: The whole function is responsible for drawing the panel. I was referring to `addon_updater_ops.update_notice_box_ui(self, context)`, which takes care of running the addon updater. Only needs to be mentioned once per addon IIRC.

File: plugin/utils/shell.py (hunk 8 above)
Comment (HENDRIX-ZT2) [reply plugin/utils/shell.py:206]: for context - the plugin itself actually has to use a hack to try and get valid bone roll for all bones. blender has a big numerical precision issue there, even if the math is perfectly correct.
so we have to verify if your math is correct or not

File: __init__.py (hunk 1 above)
Comment (HENDRIX-ZT2) [reply __init__.py:147]: Yeah, so those would go into the panel too, probably as separate operators?

Comment (SleepyEngi) [reply __init__.py:147]: changed name to "Generate Rig Edit from Pose"

File: plugin/utils/shell.py (hunk 3 above)
Comment (SleepyEngi) [reply plugin/utils/shell.py]: Converted all prints to logging.info

File: plugin/utils/shell.py (hunk 6 above)
Comment (SleepyEngi) [reply plugin/utils/shell.py]: changed to "and not p_bone.startswith("NODE_")

File: plugin/utils/shell.py (hunk 7 above)
Comment (SleepyEngi) [reply plugin/utils/shell.py]: changed to p_bone

File: plugin/utils/shell.py (hunk 4 above)
Comment (SleepyEngi) [reply plugin/utils/shell.py]: changed Mode to original_mode and Edit to editnumber

File: __init__.py (hunk 2 above)
Comment (SleepyEngi) [reply __init__.py]: removed unnecessary line

File: plugin/utils/shell.py [hunk 9]
```diff
@@ -90,22 +90,22 @@ def create_lods():
 def generate_rig_edit():
     """Automatic rig edit generator by NDP. Detects posed bones and automatically generates nodes and offsets them."""
     msgs = []
-    #logging.info(f"Generating rig edit")
-    print("-------------------------------------------------------------")
-    print("Initiating rig edit")
+    logging.info(f"Generating rig edit")
+    logging.info("-------------------------------------------------------------")
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:94]: for logging you generally don't add padding lines

File: plugin/utils/shell.py [hunk 10]
```diff
@@ -114,51 +114,51 @@ def generate_rig_edit():
...
-            print("WARNING: "+str(i.name)+" HAD SCALE, CANCELLING")
-            msgs.append(f"Warning: {str(i.name)} had scale transforms. Reset scale for all bones and try again.")
+            #p_bone.scale = Vector((1,1,1))
+            logging.info("WARNING: "+str(p_bone.name)+" HAD SCALE, CANCELLING")
+            msgs.append(f"Warning: {str(p_bone.name)} had scale transforms. Reset scale for all bones and try again.")
             return msgs
         
         #Find all bones with transforms.
-        if (i.location != Vector((0, 0, 0)) or i.rotation_quaternion != Quaternion((1,0,0,0)) or i.scale != Vector((1,1,1))) and i.name[:5] != "NODE_":
+        if (p_bone.location != Vector((0, 0, 0)) or p_bone.rotation_quaternion != Quaternion((1,0,0,0)) or p_bone.scale != Vector((1,1,1))) and not p_bone.name.startswith("NODE_"):
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:131]: I think I would separate the finding, creating and the applying into 3 separate loops?

Comment (HENDRIX-ZT2) [reply plugin/utils/shell.py:131]: if we do that, we can just use apply pose logic and just place our new nodes in the right spots

File: plugin/utils/shell.py [hunk 11]
```diff
@@ -86,6 +87,183 @@ def create_lods():
...
+    for p_bone in armature.pose.bones:
+        #Check if the bones have been scaled
+        #Update locations
+        bpy.context.view_layer.update()
+        
+        #Check that no bones are scaled
+        if p_bone.scale != Vector((1,1,1)):
+            #Reset scale
+            #p_bone.scale = Vector((1,1,1))
+            logging.info("WARNING: "+str(p_bone.name)+" HAD SCALE, CANCELLING")
```
[the remark this answers is not in the document]
Comment (HENDRIX-ZT2) [reply plugin/utils/shell.py:126]: generally use fstrings throughout

File: plugin/utils/shell.py [hunk 12]
```diff
@@ -87,17 +87,28 @@ def create_lods():
 	return msgs
 
 
+def vectorisclose(vector1,vector2,tolerance):
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:90]: make tolerance a keyword arg to avoid setting same arg in every call

File: plugin/utils/shell.py [hunk 13]
```diff
@@ -113,145 +124,194 @@ def generate_rig_edit():
...
+        #Get pose mode position of bonebase
+        bonebase_pos_pose = Vector(p_bone.matrix.translation)
+        
+        #Reset position
+        p_bone.location = Vector((0, 0, 0))
+        #Update location after moving.
+        bpy.context.view_layer.update()
+        
+        #Get rest mode position of bonebase
+        bonebase_pos_rest = Vector(p_bone.matrix.translation)
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:194]: typo - same as bonebase_pos_pose ?

File: plugin/utils/shell.py [hunk 14]
```diff
@@ -113,145 +124,194 @@ def generate_rig_edit():
...
+
+        
+        #Creating the nodes            
+        #We switch to edit mode to use edit bones, as they do not exist outside of edit mode.
+        #Switch to edit mode
+        bpy.ops.object.mode_set(mode='EDIT')
+        
+        #Get edit bones
+        bonebase = armature.data.edit_bones.get(p_bone.name)
+        boneparent = armature.data.edit_bones.get(p_bone.parent.name)
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:215]: can also do bonebase.parent

File: plugin/utils/shell.py [hunk 15]
```diff
@@ -113,145 +124,194 @@ def generate_rig_edit():
...
+        
+        #Creating the node bone
+        bonenode = armature.data.edit_bones.new(f"NODE_{p_bone.name}")
+        #NOTE: Maybe here we set the bonenode to boneparent if it starts with NODE_ already
+        
+        #Copy parent length to node
+        bonenode.length = boneparent.length
+        #Copy parent matrix to node
+        bonenode.matrix = boneparent.matrix.copy()
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:224]: Setting the matrix is known not to be reliable for bone roll. Better copy head, tail, roll

File: plugin/utils/shell.py [hunk 16]
```diff
@@ -113,145 +124,194 @@ def generate_rig_edit():
...
+        #Get edit bones
+        bonebase = armature.data.edit_bones.get(p_bone.name)
+        boneparent = armature.data.edit_bones.get(p_bone.parent.name)
+        
+        #Creating the node bone
+        bonenode = armature.data.edit_bones.new(f"NODE_{p_bone.name}")
+        #NOTE: Maybe here we set the bonenode to boneparent if it starts with NODE_ already
+        
+        #Copy parent length to node
+        bonenode.length = boneparent.length
```
Comment (HENDRIX-ZT2) [inline plugin/utils/shell.py:222]: will be redundant to set length, see below

File: plugin/utils/shell.py (hunk 13 above)
Comment (SleepyEngi) [reply plugin/utils/shell.py:194]: It's stored before and after clearing the pose. So I have the global location of the rest position and the pose position, so i can calculate the delta offset. This way i don't run into a problem where it adds up the offsets depending on the order that you calculate the offsets like with the first version of the script that caused bones to get incorrectly positioned after several chains of parents

Comment (HENDRIX-ZT2) [reply plugin/utils/shell.py:197]: ah ok I see

**Pull request merged**